content/blog/tidymodels-sparse-support/index.Rmd

```diff
@@ -0,0 +1,146 @@
+
+---
+
+
+```{r setup, include=FALSE}
+library(knitr)
+knitr::opts_chunk$set(cache = TRUE, cache.lazy = FALSE, warning = FALSE, 
+                      message = FALSE, echo = TRUE, dpi = 180,
+                      fig.width = 8, fig.height = 5)
+library(tidyverse)
+theme_set(theme_minimal())
+```
+
+The new release of [tune](linktopost) is chock full of improvements and new features. When combined with the latest releases of [hardhat](http://hardhat.tidymodels.org/) and [parsnip](https://parsnip.tidymodels.org/), one upgrade that tidymodels users can now use in their day-to-day modeling work is **support for sparse data structures**.
+
+## Why sparse data?
+
+In some subject matter domains, it is common to have lots and lots of zeroes after transforming data to a representation appropriate for analysis or modeling. Text data is one such example. The `small_fine_foods` dataset of Amazon reviews of fine foods contains a column `review` that we as humans can read and understand.
+
+```{r R.options=list(quanteda_print_dfm_max_ndoc = 0, quanteda_print_dfm_max_nfeat = 0)}
+library(tidyverse)
+library(tidymodels)
+
+data("small_fine_foods")
+training_data
+```
+
+Computers, on the other hand, need that `review` variable to be heavily preprocessed and transformed in order for it to be ready for most modeling. We typically need to [tokenize](https://smltar.com/tokenization.html) the text, find word frequencies, and perhaps [compute tf-idf](https://www.tidytextmining.com/tfidf.html). We can either keep the resulting data in a long, tidy tibble or we can transform it to a wide matrix, often a good fit when the next step is a modeling or machine learning algorithm.
+
+```{r, R.options=list(quanteda_print_dfm_max_ndoc = 0, quanteda_print_dfm_max_nfeat = 0)}
+library(tidytext)
+
+training_data %>%
+  unnest_tokens(word, review) %>%
+  count(product, word) %>%
+  bind_tf_idf(word, product, n) %>%
+  cast_dfm(product, word, tf_idf)
+```
+
+As is typical for text data, this document-feature matrix is extremely sparse, with many zeroes. Most documents do not contain most words. If we coerce this kind of structure to anything like a `data.frame`, we run into two challenges:
+
+- The amount of **memory** this object requires increases dramatically.
+- We lose out on the **speed** gained from any specialized model algorithms built for sparse data.
+
+## A blueprint for sparse models
+
+Before the most recent releases of hardhat, parsnip, and tune, there was no support for sparse data structures within tidymodels. Now, you can specify a hardhat **blueprint** for sparse data.
+
+```{r}
+library(hardhat)
+sparse_bp <- default_recipe_blueprint(composition = "dgCMatrix")
+```
+
+The `dgCMatrix` composition is from the [Matrix](https://cran.r-project.org/package=Matrix) package, and is the most standard class for sparse numeric matrices in modeling in R. You can also specify a dense matrix composition with `composition = "matrix"`.
```

Comment:
Somewhere, maybe here, we should differentiate what can _can_ do now versus what you _still can't_ do. IIRC you can start with a sparse matrix only if you use `fit_xy()` with a `parsnip` model. For example: 

```r
library(tidymodels)
library(Matrix)
library(hardhat)

data("hpc_data", package = "modeldata")

# ------------------------------------------------------------------------------

# Can do this: 
new_x <- Matrix::sparse.model.matrix(class ~ protocol + day, data = hpc_data)

cls_spec <- 
  multinom_reg(penalty = .1, mixture = 0.5) %>% 
  set_engine("glmnet", nlambda = 5)

# input is sparse
fit_xy(cls_spec, x = new_x, y = hpc_data$class)

# ------------------------------------------------------------------------------
# can do this:

sparse_bp <- default_recipe_blueprint(composition = "dgCMatrix")

rec <- 
  recipe(class ~ protocol + day, data = hpc_data) %>% 
  step_dummy(all_predictors())

wflow <- 
  workflow() %>% 
  add_recipe(rec, blueprint = sparse_bp) %>% 
  add_model(cls_spec)

#input still has to be a data frame but data gets converted and passed to model as sparse
wflow %>% fit(data = hpc_data)

# can't do anything else really
```

Somewhere we should say this.